flang/lib/Optimizer/Transforms/AffinePromotion.cpp

```diff
@@ -83,11 +101,109 @@ class AffineFunctionAnalysis {
     }
     return it->getSecond();
   }
+  AffineIfAnalysis getChildIfAnalysis(fir::WhereOp op) const {
+    auto it = ifAnalysisMap.find_as(op);
+    if (it == ifAnalysisMap.end()) {
+      LLVM_DEBUG(llvm::dbgs() << "AffineFunctionAnalysis: not computed for:\n";
+                 op.dump(););
+      op.emitError("error in fetching if analysis in AffineFunctionAnalysis\n");
+      return AffineIfAnalysis(false);
+    }
+    return it->getSecond();
+  }
   friend AffineLoopAnalysis;
+  friend AffineIfAnalysis;
 
 private:
   llvm::DenseMap<mlir::Operation *, AffineLoopAnalysis> loopAnalysisMap;
+  llvm::DenseMap<mlir::Operation *, AffineIfAnalysis> ifAnalysisMap;
 };
+
+Optional<mlir::AffineExpr> tryAffineBinaryOp(mlir::AffineExprKind kind,
+                                             Optional<mlir::AffineExpr> lhs,
+                                             Optional<mlir::AffineExpr> rhs) {
+  if (lhs.hasValue() && rhs.hasValue())
+    return mlir::getAffineBinaryOpExpr(kind, lhs.getValue(), rhs.getValue());
+  else
+    return {};
+}
+
+Optional<mlir::AffineExpr>
+toAffineExpr(mlir::Value value, llvm::SmallVector<mlir::Value, 8> &affineArgs,
+             unsigned &symPos, unsigned &dimPos) {
+  if (auto op = value.getDefiningOp<mlir::SubIOp>()) {
+    return tryAffineBinaryOp(
+        mlir::AffineExprKind::Add,
+        toAffineExpr(op.lhs(), affineArgs, symPos, dimPos),
+        tryAffineBinaryOp(
+            mlir::AffineExprKind::Mul,
+            toAffineExpr(op.rhs(), affineArgs, symPos, dimPos),
+            {mlir::getAffineConstantExpr(-1, value.getContext())}));
+  } else if (auto op = value.getDefiningOp<mlir::AddIOp>()) {
+    return tryAffineBinaryOp(
+        mlir::AffineExprKind::Add,
+        toAffineExpr(op.lhs(), affineArgs, symPos, dimPos),
+        toAffineExpr(op.rhs(), affineArgs, symPos, dimPos));
+  } else if (auto op = value.getDefiningOp<mlir::AddIOp>()) {
+    return tryAffineBinaryOp(
+        mlir::AffineExprKind::Mul,
+        toAffineExpr(op.lhs(), affineArgs, symPos, dimPos),
+        toAffineExpr(op.rhs(), affineArgs, symPos, dimPos));
+  } else if (auto op = value.getDefiningOp<mlir::UnsignedRemIOp>()) {
+    return tryAffineBinaryOp(
+        mlir::AffineExprKind::Mod,
+        toAffineExpr(op.lhs(), affineArgs, symPos, dimPos),
+        toAffineExpr(op.rhs(), affineArgs, symPos, dimPos));
```

Comment:
Some of this looks a bit repetitive. Might be able to use a template.

Reply:
I tried to reduce the repetition a bit, let me know what you think.